### PR TITLE
fx.Annotate: Support variadic functions

### PR DESCRIPTION
**Pull request by xqbumu**
Add support to fx.Annotate for variadic functions.
The last parameter of variadic functions is now treated as a slice.
With this users can feed value groups into variadic functions.

Ref https://github.com/uber-go/dig/issues/120

### CODE REVIEW

**Review by sywhang — APPROVED** [review]
Thanks for this change. Looks good to me

**Comment by xqbumu** [comment]
> 
> 
> Thanks for this change. Looks good to me

There is something wrong that 'value groups cannot be optional' when use optional tags in ParamTags.
And that message is from dig where we can delve deeper.

Refer: https://github.com/uber-go/dig/issues/120

**Comment by abhinav** [comment]
Hey, on a second look, this doesn't match the behavior of dig.
Dig specifically just ignores variadic arguments.
That is, if a function accepts variadic arguments, dig will fill everything except that.
I originally thought that's what this change was doing, but the test indicates otherwise.

If we want to support filling variadic arguments with value groups using this,
we can do that, but we should make that choice explicitly.
CC @sywhang @shirchen 

RE: value groups cannot be optional:
They cannot be optional, in that you cannot tag them optional.
They are already optional, in that if no values are provided to that group, you get an empty slice.

**Comment by xqbumu** [comment]
> 
> 
> Hey, on a second look, this doesn't match the behavior of dig. Dig specifically just ignores variadic arguments. That is, if a function accepts variadic arguments, dig will fill everything except that. I originally thought that's what this change was doing, but the test indicates otherwise.
> 
> If we want to support filling variadic arguments with value groups using this, we can do that, but we should make that choice explicitly. CC @sywhang @shirchen
> 
> RE: value groups cannot be optional: They cannot be optional, in that you cannot tag them optional. They are already optional, in that if no values are provided to that group, you get an empty slice.

Oh, I got it, very thanks for you explain.

File: annotated_test.go [hunk 1]
```diff
@@ -494,6 +498,46 @@ func TestAnnotate(t *testing.T) {
...
+
+		type T1 struct{ s string }
+
+		app := fxtest.New(t,
+			fx.Supply(
+				fx.Annotate(T1{"foo"}, fx.ResultTags(`group:"t"`)),
+				fx.Annotate(T1{"bar"}, fx.ResultTags(`group:"t"`)),
+				fx.Annotate(T1{"baz"}, fx.ResultTags(`group:"t"`)),
+			),
+			fx.Invoke(fx.Annotate(func(got []T1) {
```
Comment (sywhang) [inline annotated_test.go:532]: Should this be `got ...T1` instead?

```suggestion
			fx.Invoke(fx.Annotate(func(got ...T1) {
```

File: annotated.go [hunk 2]
```diff
@@ -522,6 +528,42 @@ func (ann *annotated) results() (
...
+//
+// You can use variadic parameters with Fx's value groups.
+// For example,
+//
+//  fx.Annotate(func(mux *http.ServeMux, handlers ...http.Handler) {
+//    // ...
+//  }, fx.ParamTags(``, `group:"server"`))
+//
+// If we provide the above to the application,
+// any constructor in the Fx application can feed it HTTP handlers
```
Comment (r-hang) [inline annotated.go:555]: nit: feed it's HTTP handlers

File: annotated_test.go (hunk 1 above)
Comment (abhinav) [reply annotated_test.go:532]: derp. yes.

File: annotated.go (hunk 2 above)
Comment (abhinav) [reply annotated.go:555]: ack

**Review by sywhang — APPROVED** [review]
LGTM

**Comment by abhinav** [comment]
Thanks! @xqbumu

**Pull request merged**